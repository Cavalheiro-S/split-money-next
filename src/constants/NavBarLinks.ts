import { Gear, House, Icon, PlusCircle, SignOut } from "@phosphor-icons/react";

export interface NavBarLink {
    name: string;
    path: string;
    icon: Icon;
    auth?: boolean;
    type: "screen" | "config"
}

export const NavBarLinks: Array<NavBarLink> = [
    {
        name: 'Visão Geral',
        path: '/',
        icon: House,
        auth: false,
        type: "screen"
    },
    {
        name: 'Lançamentos',
        path: '/transaction',
        icon: PlusCircle,
        auth: true,
        type: "screen"
    },
    {
        name: 'Sair',
        path: '/session/logout',
        icon: SignOut,
        auth: true,
        type: "config"
    }
];