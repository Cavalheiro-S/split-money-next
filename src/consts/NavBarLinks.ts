import { Gear, House, Icon, PlusCircle, SignOut } from "@phosphor-icons/react";

export interface NavBarLink {
    name: string;
    path: string;
    icon: Icon;
    type: "screen" | "config"
}

export const NavBarLinks: Array<NavBarLink> = [
    {
        name: 'Visão Geral',
        path: '/',
        icon: House,
        type: "screen"
    },
    {
        name: 'Lançamentos',
        path: '/transaction',
        icon: PlusCircle,
        type: "screen"
    },
    {
        name: 'Configurações',
        path: '/settings',
        icon: Gear,
        type: "config"
    },
    {
        name: 'Sair',
        path: '/logout',
        icon: SignOut,
        type: "config"
    }
];