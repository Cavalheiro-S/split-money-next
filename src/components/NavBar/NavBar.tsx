'use client'

import { NavBarLinks } from '@/constants/NavBarLinks';
import { RootState } from '@/store';
import { ArrowsOutLineHorizontal } from '@phosphor-icons/react';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { useState } from 'react';
import { useSelector } from 'react-redux';
import { twMerge } from 'tailwind-merge';

export const NavBar = () => {

    const router = useRouter()
    const { user } = useSelector((state: RootState) => state.userState)
    const [isNavBarOpen, setIsNavBarOpen] = useState(true)

    const renderLink = () => {
        const linksAvailables = NavBarLinks.filter(link => {
            if (link.auth && !user?.id)
                return false
            if (link.auth && user?.id)
                return true
            return true
        })
        return linksAvailables.map((linkAvailable, index) => (
            <li
                key={linkAvailable.name + index}
                className={twMerge("flex items-center gap-2", router.pathname === linkAvailable.path ? "text-gray-800" : "text-gray-500")}>
                <linkAvailable.icon className='w-6 h-6' />
                <Link href={linkAvailable.path}>{linkAvailable.name}</Link>
            </li>
        ))
    }
    return (
        <nav className={
            twMerge('absolute top-0 left-0 z-10 grid h-full grid-rows-2 pt-20 font-sans text-sm bg-white border-t-2 border-gray-200 w-60 transition-all',
                !isNavBarOpen && "-translate-x-full")}>
            <span
                onClick={() => setIsNavBarOpen(!isNavBarOpen)}
                className={twMerge('absolute -right-14 mt-0 hover:text-primary hover:cursor-pointer bg-white p-2')}>
                <ArrowsOutLineHorizontal />
                Menu
            </span>
            <ul className='flex flex-col gap-6 px-4'>
                {renderLink()}
                {!user?.id && <Link className='text-primary hover:text-primary hover:text-opacity-75' href={"/session/login"}>FaÃ§a login para acessar</Link>}
            </ul>
            <ul className='flex flex-col px-4'>
                {user?.id
                    && (<>
                        Logado como: <br />
                        <span className='font-semibold'>{`${user?.name} - ${user?.id}`}</span>
                    </>
                    )
                }
            </ul>
        </nav>
    )
}
