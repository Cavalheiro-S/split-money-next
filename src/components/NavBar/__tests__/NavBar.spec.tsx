import { render, screen } from '@testing-library/react';
import { NavBar } from '../NavBar';
import { NavBarLinks } from '@/consts/NavBarLinks';
import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
    useRouter: jest.fn(),
}));

describe('NavBar', () => {
  test('deve renderizar os links corretamente', () => {
    useRouter.mockImplementation(() => ({
      pathname: '/',
    }));

    const tree = render(<NavBar />);
    expect(tree).toMatchSnapshot();
  });

  test('deve destacar o link ativo corretamente', () => {
    useRouter.mockImplementation(() => ({
      pathname: '/transaction',
    }));

    render(<NavBar />);

    expect(screen.getByText('Lan√ßamentos').parentElement).toHaveClass('text-gray-800');
  });
});
