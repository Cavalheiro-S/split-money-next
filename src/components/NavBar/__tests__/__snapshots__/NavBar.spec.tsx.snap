// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar deve renderizar os links corretamente 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="absolute left-0 z-10 h-full pt-20 bg-white border-t-2 border-gray-200 top-20 w-60"
      >
        <ul
          class="flex flex-col gap-6 px-4"
        >
          <li
            class="flex items-center gap-2 text-gray-800"
          >
            <svg
              class="w-6 h-6"
              fill="currentColor"
              height="1em"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M218.83,103.77l-80-75.48a1.14,1.14,0,0,1-.11-.11,16,16,0,0,0-21.53,0l-.11.11L37.17,103.77A16,16,0,0,0,32,115.55V208a16,16,0,0,0,16,16H96a16,16,0,0,0,16-16V160h32v48a16,16,0,0,0,16,16h48a16,16,0,0,0,16-16V115.55A16,16,0,0,0,218.83,103.77ZM208,208H160V160a16,16,0,0,0-16-16H112a16,16,0,0,0-16,16v48H48V115.55l.11-.1L128,40l79.9,75.43.11.1Z"
              />
            </svg>
            <a
              href="/"
            >
              Visão Geral
            </a>
          </li>
          <li
            class="flex items-center gap-2 text-gray-500"
          >
            <svg
              class="w-6 h-6"
              fill="currentColor"
              height="1em"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm48-88a8,8,0,0,1-8,8H136v32a8,8,0,0,1-16,0V136H88a8,8,0,0,1,0-16h32V88a8,8,0,0,1,16,0v32h32A8,8,0,0,1,176,128Z"
              />
            </svg>
            <a
              href="/transaction"
            >
              Lançamentos
            </a>
          </li>
          <li
            class="flex items-center gap-2 text-gray-500"
          >
            <svg
              class="w-6 h-6"
              fill="currentColor"
              height="1em"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M128,80a48,48,0,1,0,48,48A48.05,48.05,0,0,0,128,80Zm0,80a32,32,0,1,1,32-32A32,32,0,0,1,128,160Zm88-29.84q.06-2.16,0-4.32l14.92-18.64a8,8,0,0,0,1.48-7.06,107.21,107.21,0,0,0-10.88-26.25,8,8,0,0,0-6-3.93l-23.72-2.64q-1.48-1.56-3-3L186,40.54a8,8,0,0,0-3.94-6,107.71,107.71,0,0,0-26.25-10.87,8,8,0,0,0-7.06,1.49L130.16,40Q128,40,125.84,40L107.2,25.11a8,8,0,0,0-7.06-1.48A107.6,107.6,0,0,0,73.89,34.51a8,8,0,0,0-3.93,6L67.32,64.27q-1.56,1.49-3,3L40.54,70a8,8,0,0,0-6,3.94,107.71,107.71,0,0,0-10.87,26.25,8,8,0,0,0,1.49,7.06L40,125.84Q40,128,40,130.16L25.11,148.8a8,8,0,0,0-1.48,7.06,107.21,107.21,0,0,0,10.88,26.25,8,8,0,0,0,6,3.93l23.72,2.64q1.49,1.56,3,3L70,215.46a8,8,0,0,0,3.94,6,107.71,107.71,0,0,0,26.25,10.87,8,8,0,0,0,7.06-1.49L125.84,216q2.16.06,4.32,0l18.64,14.92a8,8,0,0,0,7.06,1.48,107.21,107.21,0,0,0,26.25-10.88,8,8,0,0,0,3.93-6l2.64-23.72q1.56-1.48,3-3L215.46,186a8,8,0,0,0,6-3.94,107.71,107.71,0,0,0,10.87-26.25,8,8,0,0,0-1.49-7.06Zm-16.1-6.5a73.93,73.93,0,0,1,0,8.68,8,8,0,0,0,1.74,5.48l14.19,17.73a91.57,91.57,0,0,1-6.23,15L187,173.11a8,8,0,0,0-5.1,2.64,74.11,74.11,0,0,1-6.14,6.14,8,8,0,0,0-2.64,5.1l-2.51,22.58a91.32,91.32,0,0,1-15,6.23l-17.74-14.19a8,8,0,0,0-5-1.75h-.48a73.93,73.93,0,0,1-8.68,0,8,8,0,0,0-5.48,1.74L100.45,215.8a91.57,91.57,0,0,1-15-6.23L82.89,187a8,8,0,0,0-2.64-5.1,74.11,74.11,0,0,1-6.14-6.14,8,8,0,0,0-5.1-2.64L46.43,170.6a91.32,91.32,0,0,1-6.23-15l14.19-17.74a8,8,0,0,0,1.74-5.48,73.93,73.93,0,0,1,0-8.68,8,8,0,0,0-1.74-5.48L40.2,100.45a91.57,91.57,0,0,1,6.23-15L69,82.89a8,8,0,0,0,5.1-2.64,74.11,74.11,0,0,1,6.14-6.14A8,8,0,0,0,82.89,69L85.4,46.43a91.32,91.32,0,0,1,15-6.23l17.74,14.19a8,8,0,0,0,5.48,1.74,73.93,73.93,0,0,1,8.68,0,8,8,0,0,0,5.48-1.74L155.55,40.2a91.57,91.57,0,0,1,15,6.23L173.11,69a8,8,0,0,0,2.64,5.1,74.11,74.11,0,0,1,6.14,6.14,8,8,0,0,0,5.1,2.64l22.58,2.51a91.32,91.32,0,0,1,6.23,15l-14.19,17.74A8,8,0,0,0,199.87,123.66Z"
              />
            </svg>
            <a
              href="/settings"
            >
              Configurações
            </a>
          </li>
          <li
            class="flex items-center gap-2 text-gray-500"
          >
            <svg
              class="w-6 h-6"
              fill="currentColor"
              height="1em"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M112,216a8,8,0,0,1-8,8H48a16,16,0,0,1-16-16V48A16,16,0,0,1,48,32h56a8,8,0,0,1,0,16H48V208h56A8,8,0,0,1,112,216Zm109.66-93.66-40-40a8,8,0,0,0-11.32,11.32L196.69,120H104a8,8,0,0,0,0,16h92.69l-26.35,26.34a8,8,0,0,0,11.32,11.32l40-40A8,8,0,0,0,221.66,122.34Z"
              />
            </svg>
            <a
              href="/logout"
            >
              Sair
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="absolute left-0 z-10 h-full pt-20 bg-white border-t-2 border-gray-200 top-20 w-60"
    >
      <ul
        class="flex flex-col gap-6 px-4"
      >
        <li
          class="flex items-center gap-2 text-gray-800"
        >
          <svg
            class="w-6 h-6"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M218.83,103.77l-80-75.48a1.14,1.14,0,0,1-.11-.11,16,16,0,0,0-21.53,0l-.11.11L37.17,103.77A16,16,0,0,0,32,115.55V208a16,16,0,0,0,16,16H96a16,16,0,0,0,16-16V160h32v48a16,16,0,0,0,16,16h48a16,16,0,0,0,16-16V115.55A16,16,0,0,0,218.83,103.77ZM208,208H160V160a16,16,0,0,0-16-16H112a16,16,0,0,0-16,16v48H48V115.55l.11-.1L128,40l79.9,75.43.11.1Z"
            />
          </svg>
          <a
            href="/"
          >
            Visão Geral
          </a>
        </li>
        <li
          class="flex items-center gap-2 text-gray-500"
        >
          <svg
            class="w-6 h-6"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm48-88a8,8,0,0,1-8,8H136v32a8,8,0,0,1-16,0V136H88a8,8,0,0,1,0-16h32V88a8,8,0,0,1,16,0v32h32A8,8,0,0,1,176,128Z"
            />
          </svg>
          <a
            href="/transaction"
          >
            Lançamentos
          </a>
        </li>
        <li
          class="flex items-center gap-2 text-gray-500"
        >
          <svg
            class="w-6 h-6"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128,80a48,48,0,1,0,48,48A48.05,48.05,0,0,0,128,80Zm0,80a32,32,0,1,1,32-32A32,32,0,0,1,128,160Zm88-29.84q.06-2.16,0-4.32l14.92-18.64a8,8,0,0,0,1.48-7.06,107.21,107.21,0,0,0-10.88-26.25,8,8,0,0,0-6-3.93l-23.72-2.64q-1.48-1.56-3-3L186,40.54a8,8,0,0,0-3.94-6,107.71,107.71,0,0,0-26.25-10.87,8,8,0,0,0-7.06,1.49L130.16,40Q128,40,125.84,40L107.2,25.11a8,8,0,0,0-7.06-1.48A107.6,107.6,0,0,0,73.89,34.51a8,8,0,0,0-3.93,6L67.32,64.27q-1.56,1.49-3,3L40.54,70a8,8,0,0,0-6,3.94,107.71,107.71,0,0,0-10.87,26.25,8,8,0,0,0,1.49,7.06L40,125.84Q40,128,40,130.16L25.11,148.8a8,8,0,0,0-1.48,7.06,107.21,107.21,0,0,0,10.88,26.25,8,8,0,0,0,6,3.93l23.72,2.64q1.49,1.56,3,3L70,215.46a8,8,0,0,0,3.94,6,107.71,107.71,0,0,0,26.25,10.87,8,8,0,0,0,7.06-1.49L125.84,216q2.16.06,4.32,0l18.64,14.92a8,8,0,0,0,7.06,1.48,107.21,107.21,0,0,0,26.25-10.88,8,8,0,0,0,3.93-6l2.64-23.72q1.56-1.48,3-3L215.46,186a8,8,0,0,0,6-3.94,107.71,107.71,0,0,0,10.87-26.25,8,8,0,0,0-1.49-7.06Zm-16.1-6.5a73.93,73.93,0,0,1,0,8.68,8,8,0,0,0,1.74,5.48l14.19,17.73a91.57,91.57,0,0,1-6.23,15L187,173.11a8,8,0,0,0-5.1,2.64,74.11,74.11,0,0,1-6.14,6.14,8,8,0,0,0-2.64,5.1l-2.51,22.58a91.32,91.32,0,0,1-15,6.23l-17.74-14.19a8,8,0,0,0-5-1.75h-.48a73.93,73.93,0,0,1-8.68,0,8,8,0,0,0-5.48,1.74L100.45,215.8a91.57,91.57,0,0,1-15-6.23L82.89,187a8,8,0,0,0-2.64-5.1,74.11,74.11,0,0,1-6.14-6.14,8,8,0,0,0-5.1-2.64L46.43,170.6a91.32,91.32,0,0,1-6.23-15l14.19-17.74a8,8,0,0,0,1.74-5.48,73.93,73.93,0,0,1,0-8.68,8,8,0,0,0-1.74-5.48L40.2,100.45a91.57,91.57,0,0,1,6.23-15L69,82.89a8,8,0,0,0,5.1-2.64,74.11,74.11,0,0,1,6.14-6.14A8,8,0,0,0,82.89,69L85.4,46.43a91.32,91.32,0,0,1,15-6.23l17.74,14.19a8,8,0,0,0,5.48,1.74,73.93,73.93,0,0,1,8.68,0,8,8,0,0,0,5.48-1.74L155.55,40.2a91.57,91.57,0,0,1,15,6.23L173.11,69a8,8,0,0,0,2.64,5.1,74.11,74.11,0,0,1,6.14,6.14,8,8,0,0,0,5.1,2.64l22.58,2.51a91.32,91.32,0,0,1,6.23,15l-14.19,17.74A8,8,0,0,0,199.87,123.66Z"
            />
          </svg>
          <a
            href="/settings"
          >
            Configurações
          </a>
        </li>
        <li
          class="flex items-center gap-2 text-gray-500"
        >
          <svg
            class="w-6 h-6"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M112,216a8,8,0,0,1-8,8H48a16,16,0,0,1-16-16V48A16,16,0,0,1,48,32h56a8,8,0,0,1,0,16H48V208h56A8,8,0,0,1,112,216Zm109.66-93.66-40-40a8,8,0,0,0-11.32,11.32L196.69,120H104a8,8,0,0,0,0,16h92.69l-26.35,26.34a8,8,0,0,0,11.32,11.32l40-40A8,8,0,0,0,221.66,122.34Z"
            />
          </svg>
          <a
            href="/logout"
          >
            Sair
          </a>
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
